api/pin.go

```diff
@@ -0,0 +1,426 @@
+// that are encountered on the way. The pre-requisite is that the file should
+// have been already pinned using the PinFiles function. This function can
+// be called only from an external command.
+func (p *PinAPI) UnpinFiles(rootHash string, credentials string) {
+
+	addr, err := hex.DecodeString(rootHash)
+	if err != nil {
+		log.Error("Error decoding root hash" + err.Error())
+		return
+	}
+
+	isRawInDB, err := p.db.IsPinnedFileRaw(chunk.Address(addr))
+	if err != nil {
+		log.Error("Root hash is not pinned" + err.Error())
+		return
+	}
+
+	// Walk the root hash and unpin all the chunks
+	walkerFunction := func(ref storage.Reference) error {
+		chunkAddr := p.removeDecryptionKeyFromChunkHash(ref)
+		err := p.db.Set(context.TODO(), chunk.ModeSetUnpin, chunkAddr)
+		if err != nil {
+			log.Error("Could not unpin chunk. Address " + hex.EncodeToString(chunkAddr))
+			return err
+		} else {
+			log.Trace("Unpinning chunk", "Address", fmt.Sprintf("%0x", chunkAddr))
+		}
+		return nil
+	}
+	p.walkChunksFromRootHash(rootHash, isRawInDB, credentials, walkerFunction)
+
+	// Check if the root chunk exists in pinIndex
+	// If it is not.. then the pin counter became 0
+	// so remove the root hash from pinFilesIndex
+	isRootChunkPinned := p.db.IsChunkPinned(chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+	if !isRootChunkPinned {
+		err := p.db.Set(context.TODO(), chunk.ModeSetUnpinFile, addr)
+		if err != nil {
+			// TODO: if this happens, we should go back and revert the entire file's chunks
+			log.Error("Could not unpin root chunk. Address " + fmt.Sprintf("%0x", addr))
+		}
+	}
+}
+
+// ListPinFiles functions logs information of all the files that are pinned
+// in the current local node. It displays the root hash of the pinned file
+// or collection. It also display two vital information
+//     1) Size of the pinned file or collection
+//     2) the number of times that particular file or collection is pinned.
+func (p *PinAPI) ListPinFiles() {
+	pinnedFiles := p.db.GetPinFilesIndex()
+	for k, v := range pinnedFiles {
+
+		addr, err := hex.DecodeString(k)
+		if err != nil {
+			log.Error("Error decoding root hash" + err.Error())
+			return
+		}
+		pinCounter, err := p.db.GetPinCounterOfChunk(chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+
+		isRaw := false
+		if v > 0 {
+			isRaw = true
+		}
+		noOfChunks := p.getNoOfChunks(k, isRaw, "")
+
+		log.Info("Pinned file", "Address", k, "NoOfChunks", noOfChunks, "pinCounter", pinCounter)
+	}
+}
+
+// LogPinnedChunks logs all the chunks that are pinned as part of the given root hash.
+// This is mostly used for debugging.
+func (p *PinAPI) LogPinnedChunks(rootHash string, credentials string) {
+
+	addr, err := hex.DecodeString(rootHash)
+	if err != nil {
+		log.Error("Error decoding root hash" + err.Error())
+		return
+	}
+
+	isRawInDB, err := p.db.IsPinnedFileRaw(chunk.Address(addr))
+	if err != nil {
+		log.Error("Root hash is not pinned" + err.Error())
```

Comment:
function should return an error and propagated to calling context

Reply:
Done